### PR TITLE
New new qt layerlist

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
pulled out from and closes #2423.  This PR swaps out the old `QtLayerlist` (derived from `QScrollArea` with individual `QLayerListWidgets` for each layer in the "list")... for a new `QtLayerlist` that is just a tiny modification of the generic `QtListView` introduced in #2486, along with a custom `LayerDelegate` that controls the visual appearance of each layer in the view.

This is the PR that (finally) removes a lot of code :) diff relative to 2486 [here](https://github.com/tlambert03/napari/compare/qtlistview...tlambert03:new-new-qt-layerlist)

still need to update tests...

## Type of change

- [x] Refactor

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by tlambert03** [comment]
OK, I'd say this is ready for people to pull, play with, and review.  Shouldn't be anything breaking here from the user perspective, but it pulls out a pretty large custom component (the `QtLayerList`) and uses a very thin subclass of the generic `QItemView` instead.  The nice thing here is that almost no additional custom logic is required on the view (over the generic list view).  All of the "layer-specific" stuff comes from the [`LayerDelegate`](https://github.com/napari/napari/pull/2493/files#diff-eeac0e32eb670a45245c01e4d5ad46e711a00bed172205dc195818bedd177e77R59) which provides a little additional custom styling and logic for drawing the thumbnails, and the [`QtLayerListModel`](https://github.com/napari/napari/pull/2493/files#diff-951ca903b854af73d46597275644d6b6d35a0f5da5137979e7cf409b833ceb19R8), which knows how to provide the view/delegate with specific data "roles" (e.g. layer name, whether it's visible, the thumbnail, etc...)

This general pattern can now be reused by anybody creating a `SelectableEventedList` of any custom Python object.  They will just need to update `model.data()` and `model.setData()` for the attributes on their object type, and (optionally) customize the delegate if they want to stylize.  @emilmelnikov, this is what I was referring to in [the zulip thread the other day](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/Qt.20models.20for.20core.20Napari.20types/near/232289554), curious to hear if this meets your needs.

**Comment by tlambert03** [comment]
bonus, the new list looks a little sleeker, particularly when it comes to drag/operations! 🎉

https://user-images.githubusercontent.com/1609449/113213455-3933e380-9246-11eb-87c7-8b11f8aabdd6.mov

File: napari/components/_viewer_key_bindings.py [hunk 1]
```diff
@@ -134,40 +134,40 @@ def remove_all_layers(viewer):
...
-    """Also select layer below."""
-    viewer.layers.select_previous(shift=True)
+# @ViewerModel.bind_key('Shift-Down')
+# def also_select_layer_below(viewer):
+#     """Also select layer below."""
+#     viewer.layers.select_previous(shift=True)
 
 
-also_select_layer_below.__doc__ = trans._("Also select layer below.")
+# also_select_layer_below.__doc__ = trans._("Also select layer below.")
```
Comment (Carreau) [inline napari/components/_viewer_key_bindings.py]: Forgot to uncomment code ?

File: napari/_qt/containers/_layer_delegate.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        editor = super().createEditor(parent, option, index)
+        # make sure editor has same alignment as the display name
+        editor.setAlignment(Qt.Alignment(index.data(Qt.TextAlignmentRole)))
+        return editor
+
+    def editorEvent(
+        self,
+        event: QtCore.QEvent,
+        model: QtCore.QAbstractItemModel,
+        option: 'QStyleOptionViewItem',
```
Comment (Carreau) [inline napari/_qt/containers/_layer_delegate.py]:
```suggestion
        option: QStyleOptionViewItem,
```

**Review by Carreau — APPROVED** [review]

File: napari/components/_viewer_key_bindings.py (hunk 1 above)
Comment (tlambert03) [reply napari/components/_viewer_key_bindings.py]: oh right, thanks for reminding me! Yeah we need to discuss this.  All of these keybindings are [provided by default by the QItemView](https://doc.qt.io/qt-5/qabstractitemview.html#details).  We likely originally had all this there because we implemented the LayerList from scratch, but now, technically, we don't need any of it anymore.  _However_: if we want, we can keep it all here and probably add logic to disable it on the QtView ... something to discuss

File: napari/_qt/containers/_layer_delegate.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+if TYPE_CHECKING:
+    from qtpy.QtCore import QModelIndex
+    from qtpy.QtGui import QPainter
+    from qtpy.QtWidgets import QStyleOptionViewItem, QWidget
+
+# translations from the old qt layer list
+# tb.setToolTip(trans._('Layer thumbnail'))
+# cb.setToolTip(trans._('Layer visibility'))
+# ltb.setToolTip(trans._('Layer type'))
+# msg = trans._('Click to select\nDrag to rearrange')
```
Comment (goanpeca) [inline napari/_qt/containers/_layer_delegate.py]: What about these comments?

File: napari/components/_viewer_key_bindings.py (hunk 1 above)
Comment (goanpeca) [reply napari/components/_viewer_key_bindings.py]: Widgets that provide `motion` shortcuts by default are usually generic and general enough, that we should not allow for customization of shortcuts, plus less code to maintain, so I would go with removing them.

**Review by goanpeca — APPROVED** [review]
Awesome work @tlambert03 !

File: napari/components/_viewer_key_bindings.py (hunk 1 above)
Comment (tlambert03) [reply napari/components/_viewer_key_bindings.py]: the main issue here is the scope/context of the shortcuts.  The behavior that everyone is currently used (for better or worse) to is that up/down will work even if the layerlist is not focused.  For instance, press "add new points" button a few times, and then use up/down (it changes the layer).  If we use the builtins, then we either need to decide we don't want that behavior (consciously) or make sure to pass certain keypresses from the window down to the layerlist

Comment (sofroniewn) [reply napari/components/_viewer_key_bindings.py]: i think probably wise to drop, i like the global scope, but probably not worth the custom code/ complexity. if we had been there from the beginning we never would have added

Comment (tlambert03) [reply napari/components/_viewer_key_bindings.py]: great. I'd like let Qt handle the viewer motion shortcuts too, as @goanpeca preferred.  we can address desired behavior via Qt in another PR

File: napari/_qt/containers/_layer_delegate.py (hunk 3 above)
Comment (tlambert03) [reply napari/_qt/containers/_layer_delegate.py]: removed

File: napari/components/_viewer_key_bindings.py (hunk 1 above)
Comment (tlambert03) [reply napari/components/_viewer_key_bindings.py]: I removed all the unneeded bindings.  To get the global scope back, I _think_ we just need to pass keyPressEvent from qt_main_window to qtlayerlist (@goanpeca, does that sound right?)

File: napari/_qt/containers/_layer_delegate.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+from ..qt_resources import QColoredSVGIcon
+
+if TYPE_CHECKING:
+    from qtpy.QtCore import QModelIndex
+    from qtpy.QtGui import QPainter
+    from qtpy.QtWidgets import QStyleOptionViewItem, QWidget
+
+
+class LayerDelegate(QStyledItemDelegate):
```
Comment (sofroniewn) [inline napari/_qt/containers/_layer_delegate.py:55]: Not sure if done here or in stylesheet, but can we make the height of each of the items a little less, they seem a bit taller to me than before, but in general having them thinner will make more visible at any moment. Happy to have that in a followup PR, but noting now

File: napari/_qt/qt_viewer.py [hunk 5]
```diff
@@ -220,6 +220,17 @@ def __init__(self, viewer: Viewer, welcome=False):
             self.viewer.layers, self._qt_poll
         )
 
+        # moved from the old layerlist... still feels misplaced.
```
Comment (sofroniewn) [inline napari/_qt/qt_viewer.py:223]: Let's do in dedicated PR, but thanks for noting

**Review by sofroniewn — APPROVED** [review]
Love this @tlambert03 so great to have all that highly custom complex code go. I'm approving this looks good to go by me. I'm fine with losing global scope of up/down keybindings.

I requested that the individual items be a little thinner, it looks like they got a bit thicker to me, but even if not in general having them thinner would probably be good so you can see more at the same time. Not critical though.

Maybe @jni wants to take a look now too, this PR is the culmination of quite a few!!

**Comment by jni** [comment]
Um, layer ordering appears to be reversed?

Screen Shot 2021-04-06 at 2 29 34 pm

**Comment by tlambert03** [comment]
Oh good catch, forgot that we show things at the end of the list first (that order actually reflects the python object, which is reversed from the display order)

File: napari/_qt/containers/qt_layer_model.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        if role == Qt.DisplayRole:  # used for item text
+            return layer.name
+        if role == Qt.TextAlignmentRole:  # alignment of the text
+            return Qt.AlignCenter
+        if role == Qt.EditRole:  # used to populate line edit when editing
+            return layer.name
+        if role == Qt.ToolTipRole:  # for tooltip
+            return layer.name
+        if role == Qt.CheckStateRole:  # the "checked" state of this item
+            return Qt.Checked if layer.visible else Qt.Unchecked
```
Comment (emilmelnikov) [inline napari/_qt/containers/qt_layer_model.py:23]: I've experimented with this branch, and it works, thanks @tlambert03!

I needed to work around `QtLayerListModel` connected to a combo box though: model's `CheckStateRole` for a particular row depends on layer visibility for that row, which clashes with combo box check state (all items appear to be checked). I overrided `data()` to return non-null values on "relevant" roles, not sure if this is the right approach.

File: napari/components/_viewer_key_bindings.py (hunk 1 above)
Comment (goanpeca) [reply napari/components/_viewer_key_bindings.py]: >(@goanpeca, does that sound right?)

It does :)

**Comment by tlambert03** [comment]
> Um, layer ordering appears to be reversed?

ok @jni.  I reverse the view order.  ready for your review again

File: napari/_qt/containers/qt_layer_model.py (hunk 6 above)
Comment (tlambert03) [reply napari/_qt/containers/qt_layer_model.py:23]: >  I overrided data() to return non-null values on "relevant" roles, not sure if this is the right approach.

i'd probably need to see a bit more code (not sure I'm following what you mean about the combo box here) ... but overriding `data()` to customize stuff for your needs is absolutely a good/standard approach.  So no worries there!

Comment (emilmelnikov) [reply napari/_qt/containers/qt_layer_model.py:23]: > i'd probably need to see a bit more code (not sure I'm following what you mean about the combo box here)

```python
box = QComboBox()
model = napari_viewer.window.qt_viewer.layers.model()
box.setModel(model)
```

image

I find it strange because [QComboBox doc](https://doc.qt.io/qt-5/qcombobox.html) says:

> For the text and icon in the combobox label, the data in the model that has the Qt::DisplayRole and Qt::DecorationRole is used.

**Comment by sofroniewn** [comment]
@tlambert03 we're getting some consistent CI fails here, not sure why, but i tried rerunning and they failed again

File: napari/_qt/containers/_base_item_model.py [hunk 7]
```diff
@@ -14,6 +14,9 @@
 
 ItemType = TypeVar("ItemType")
 
+ItemRole = Qt.UserRole
+SortRole = Qt.UserRole + 1
```
Comment (jni) [inline napari/_qt/containers/_base_item_model.py:18]: Not something that I want to hold up this PR (especially since it's not *really* new code), but my English -> unfamiliar code parser is breaking here. Can you explain what these mean? Possibly a link to relevant docs? My quick Google search only took me to confused users and unhelpful responses. :joy:

Comment (goanpeca) [reply napari/_qt/containers/_base_item_model.py:18]: See https://doc.qt.io/qt-5/qt.html#ItemDataRole-enum

Screen Shot 2021-04-13 at 6 41 15 PM

It seems `SortRole` is a custom role we are defining, so we need to use the next available number, that is `Qt.UserRole + 1`

File: napari/_qt/containers/_layer_delegate.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+   but see ``QCommonStyle`` for a generic implementation)
+    a. It is also possible to override the view's style, using either a
+    subclass of ``QCommonStyle``, for a platform-independent look and feel, or
+    ``QProxyStyle``, which let's you override only certain stylistic elements
+    on any platform, falling back to the system default otherwise.
+    b. ``QStyle`` paints various elements using methods like ``drawPrimitive``
+       and ``drawControl``.  These can be overridden for very fine control.
+6. It is hard to use stylesheets with custom ``QStyles``... but it's possible
+   to style sub-controls in ``QAbstractItemView`` (such as ``QTreeView``):
+   https://doc.qt.io/qt-5/stylesheet-reference.html#list-of-sub-controls
```
Comment (jni) [inline napari/_qt/containers/_layer_delegate.py:33]: :+1: !

**Review by jni — APPROVED** [review]
looks and works great, thanks @tlambert03!

File: napari/_qt/containers/_base_item_model.py (hunk 7 above)
Comment (jni) [reply napari/_qt/containers/_base_item_model.py:18]: That definitely clarifies one half of it, thanks @goanpeca! But I don't understand the very concept of "roles"...

Comment (tlambert03) [reply napari/_qt/containers/_base_item_model.py:18]: I can add an extended docstring somewhere to explain it... since it's kind of important for anyone who would want to subclass this for something other than the layer list.  but the main idea is that when you try to get information about an item in the model, rather than just getting the item itself (like `item = model[index]`), you ask for certain "roles" like `text = model.data(index, Qt.DisplayRole)` or `icon = model.data(index, Qt.DecorationRole)` or `visible = model.data(index, Qt.CheckedRole)`, etc... [Here's the key part with the layerlist](https://github.com/tlambert03/napari/blob/ac6427fdfae4bc29b162c51901d76ab0b86ccfd0/napari/_qt/containers/qt_layer_model.py#L11-L38).  Most of those roles have builtin-meaning: like Qt builds the bulk of the actual visual appearance of each "widget" in the layer list based off of those little "roles" of data (though they're not actually QWidgets in this case)... but you can also create your own roles in certain cases.
in summary: the `model.data()` method is the heart of the translation between the object that is actually _in_ the model, and anything else (like a view) that wants to present some information about the items in a model
